bundles/org.openhab.binding.verisure/src/main/java/org/openhab/binding/verisure/internal/discovery/VerisureThingDiscoveryService.java

```diff
@@ -0,0 +1,177 @@
+
+    private @NonNullByDefault({}) VerisureBridgeHandler verisureBridgeHandler;
+    private @NonNullByDefault({}) ThingUID bridgeUID;
+
+    public VerisureThingDiscoveryService() {
+        super(VerisureHandlerFactory.SUPPORTED_THING_TYPES, SEARCH_TIME_SECONDS);
+    }
+
+    @Override
+    public void startScan() {
+        removeOlderResults(getTimestampOfLastScan());
+        logger.debug("VerisureThingDiscoveryService:startScan");
+        if (verisureBridgeHandler != null) {
+            VerisureSession session = verisureBridgeHandler.getSession();
+            if (session != null) {
+                HashMap<String, VerisureThing> verisureThings = session.getVerisureThings();
+                for (Map.Entry<String, VerisureThing> entry : verisureThings.entrySet()) {
+                    VerisureThing thing = entry.getValue();
+                    if (thing != null) {
+                        logger.debug("Thing: {}", thing);
+                        onThingAddedInternal(thing);
+                    }
+                }
+            }
+        }
+    }
+
+    private void onThingAddedInternal(VerisureThing thing) {
+        logger.debug("VerisureThingDiscoveryService:OnThingAddedInternal");
+        ThingUID thingUID = getThingUID(thing);
+        String deviceId = thing.getDeviceId();
+        if (thingUID != null && deviceId != null) {
+            if (verisureBridgeHandler != null) {
+                String label = "Device Id: " + deviceId;
+                if (thing.getLocation() != null) {
+                    label += ", Location: " + thing.getLocation();
+                }
+                if (thing.getSiteName() != null) {
+                    label += ", Site name: " + thing.getSiteName();
+                }
+                DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(thingUID).withBridge(bridgeUID)
+                        .withLabel(label).withProperty(VerisureThingConfiguration.DEVICE_ID_LABEL, deviceId).build();
+                logger.debug("thinguid: {}, bridge {}, label {}", thingUID.toString(), bridgeUID, thing.getDeviceId());
+                thingDiscovered(discoveryResult);
+            }
+        } else {
+            logger.debug("Discovered unsupported thing of type '{}' with deviceId {}", thing.getClass(),
+                    thing.getDeviceId());
+        }
+
+    }
+
+    private @Nullable ThingUID getThingUID(VerisureThing thing) {
+        ThingUID thingUID = null;
+        if (verisureBridgeHandler != null) {
+            String deviceId = thing.getDeviceId();
+            if (deviceId != null) {
+                // Make sure device id is normalized, i.e. replace all non character/digits with empty string
+                deviceId.replaceAll("[^a-zA-Z0-9]+", "");
+                if (thing instanceof VerisureAlarms) {
+                    thingUID = new ThingUID(THING_TYPE_ALARM, bridgeUID, deviceId);
```

Comment:
I'm still thinking about this comment, last before I'll push my changes. I think the pattern that I've used to use instanceof is a common one but I think I understand what you to achieve and that sounds like a better solution since the binding uses spo many different thing types. I think I would like to try a solution for this before pushing the rest of the changes for your very detailed and valid review comments.

Reply:
I think your suggested solution ended up much better and resulting in cleaner solution! :-)
Ill let my latest jar run for a while in my production environment before pushing tomorrow.

Reply:
@cpmeister I've now pushed a lot of updated files after your extremly accurate and in depth going code review, I must say that I'm impressed with your thorough review, especially since this binding has been under review for over one year. :-) I've learnt a lot from it.

I've tried to fix as many of the compilation warnings as possible, however some of the warnings are not correct so I see no other way than letting them be.